spring4/src/main/java/org/jdbi/v3/spring4/JdbiTemplate.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.spring4;
+
+import org.jdbi.v3.core.Handle;
+import org.jdbi.v3.core.HandleCallback;
+import org.jdbi.v3.core.Jdbi;
+
+/**
+ * Spring template for querying with Jdbi. This class is transaction aware and respects spring's
+ * {@link org.springframework.transaction.annotation.Transactional}
+ */
+public class JdbiTemplate implements JdbiOperations {
+
+    private final Jdbi jdbi;
+
+    public JdbiTemplate(Jdbi jdbi) {
+        this.jdbi = jdbi;
+    }
+
+    /**
+     * like {@link Jdbi#withHandle(HandleCallback)} but respects springs transactions
+     */
+    @Override
```

Comment:
Ok, I think I am starting to see the intent here.  But I wonder, it looks like the only purpose of `JdbiTemplate` is to modify the behavior of `withHandle`.  Instead of introducing it as a wrapper with an interface, what if we wrap the `ConnectionFactory` that gets called in `Jdbi.open()` to invoke the `JdbiUtil.getHandle`?  And for the close logic, you could hook the `StatementBuilder.close()` logic

Then there is no need for this wrapper class, I think.  Does that sound like it might work?

Reply:
Interesting, let me think this through loudly:

If the connectionfactory is "spring transaction aware" the callstack of using jdbi within a spring transaction would approximately look like this

`Jdbi.useHandle` -> `JdbiUtil.getHandle` -> doStuffWith handle -> `Handle.close` -> `StatementBuilder.close` and `Connection.close`

everything after the "do Stuff" I find problematic. 
- The handle does not know anything about about the transaction
- `StatementBuilder.close` could do the right thing, thats okay!
- the `java.sql.Connection` is closed (Line 132) - bad!
- the handle is marked as closed (released) which may or may not be correct!
- ...

also:
- what if you used `Jdbi.open directly` - you now must not close it but use `JdbiUtil.closeIfNeeded`. are you supposed to know that?
- what if you create a transaction manually? `Jdbi.useTransaction` and so on
...

I have had these concerns before and thats why I decided 
- to wrap the calls to Jdbi via transaction aware classes - so Jdbi is properly informed about every change in the transactions and the handle state reflects the state in the transaction.
- to discourage the use of transaction methods on `Jdbi` - you can start a transaction programmatically via springs mechanisms
- to expose a spring-like API for convenience

If you see any possibility to go without the API  - I am happy to implement it that way. But I am not really happy with static calls and Jdbi not knowing about springs state and vice versa